### ISSUE TITLE
Score2Perf and relative positional encoding

### ISSUE THREAD

**Issue by vidavakil** [tensorflow/magenta#1486]
Hello,

Reporting a possible issue with Score2Perf and tensor2tensor.

Based on its blog post, my understanding is that Music Transformer uses self_attention_type = dot_product_relative_v2. 
Looking at the implementation of dot_product_self_attention_relative_v2() in tensor2tensor codebase, it sounds as if the intention was to only look at past tokens (to the left of any token). However, there are multiple lines in the code (see below) that allow a token to look at future tokens, because of the absence of proper masking. I have marked these lines with a numbered comment .

    logits = tf.matmul(q, k, transpose_b=True)    # 1
    key_relative_embeddings = get_relative_embeddings_left(
        max_relative_position, length, depth_k, num_heads,
        heads_share_relative_embedding, "key_relative_embeddings")

    rel_logits = matmul_with_relative_keys(q, key_relative_embeddings,
                                           heads_share_relative_embedding)  # 2 
    rel_logits = _relative_position_to_absolute_position_masked(rel_logits)
    logits += rel_logits     # 3
    if bias is not None:
      logits += bias    # 4

    weights = tf.nn.softmax(logits, name="attention_weights")

Shouldn't the upper triangle of logits be replaced with very small negative numbers before being passed to softmax? I do see that kind of logic used in other types of attention in the tensor2tensor code (e.g., using attention_bias_lower_triangle()) , but not in the above code.

Meanwhile, I am training a model with the original code above: 50K steps of training so far, eval NLL quite higher than reported in the paper, generated performances a bit too much structured and taking a very long time to generate on a new Mac. My main concern is that in the absence of masking in the above code, the eval NLL would not be accurate, because while during training/eval tokens pay attention to all future tokens in the complete sequence, that is not the case during generation, because of the autoregressive nature of generation (though, with the above code, even during generation each token will pay attention to all future tokens generated so far that are ahead of it). In the absence of masking attention to future tokens, the correct way of measuring eval NLL would be to run the model auto-regressively. And yet, the model would not be trained correctly anyway, because it won't be causal. Is your paper/results based on a different/corrected implementation of above code? Or is the absence of masking intentional, because in spite of the technical issues the result still works rather well? If so, was the reported eval NLL correctly measured?

Would appreciate hearing from you, and about your plans to update/release code for Music Transformer.

Thanks.

**Comment by vidavakil** [comment]
Please ignore the above! I had missed that the passed in bias doubles as the mask and is supposedly lower-diagonal.

Best.
